libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -978,8 +980,30 @@ void IRGeneratorForStatements::endVisit(IndexAccess const& _indexAccess)
 			}
 			case DataLocation::CallData:
 			{
-				solUnimplemented("calldata not yet implemented!");
-
+				vector<string> vars { m_context.newYulVariable() };
+				if (arrayType.baseType()->isDynamicallySized())
+					vars.emplace_back(m_context.newYulVariable());
+				m_code << "let " <<
+					joinHumanReadable(vars) <<
+					" := " <<
+					m_utils.calldataArrayIndexAccessFunction(arrayType) <<
+					"(" <<
+					m_context.variable(_indexAccess.baseExpression()) <<
+					", " <<
+					expressionAsType(*_indexAccess.indexExpression(), *TypeProvider::uint256()) <<
+					")\n";
+				if (arrayType.isByteArray())
+					defineExpression(_indexAccess) << m_utils.conversionFunction(*arrayType.baseType(), *arrayType.baseType()) << "(calldataload(" << vars.front() << "))\n";
+				else if (arrayType.baseType()->isValueType())
+					defineExpression(_indexAccess) << m_utils.readFromCalldata(*arrayType.baseType()) << "(" << vars.front() << ")\n";
+				else if (arrayType.baseType()->sizeOnStack() > 1)
```

Comment:
@chriseth these here are the ones I really don't like - this is the workarounds for not having tuple assignments... (if we had them, we could just always use ``defineExpression(_indexAccess) << joinHumanReadable(vars) << "\n";)`` in all cases like this)

Reply:
`defineExpression(_indexAccess, {"offset", "length"}, vars)`